packages/flutter_tools/lib/src/runner/flutter_command_runner.dart

```diff
@@ -409,25 +409,27 @@ class FlutterCommandRunner extends CommandRunner<void> {
 
     if (engineSourcePath == null && globalResults['local-engine'] != null) {
       try {
-        final Uri engineUri = PackageMap(PackageMap.globalPackagesPath).map[kFlutterEnginePackageName];
+        Uri engineUri = PackageMap(PackageMap.globalPackagesPath).map[kFlutterEnginePackageName];
+        // Skip if sky_engine is the self-contained one.
+        if (engineUri?.path == fs.path.join(Cache.flutterRoot, 'bin', 'cache', 'pkg', kFlutterEnginePackageName, 'lib') + fs.path.separator) {
+          engineUri = null;
+        }
+        // If sky_engine is specified and the engineSourcePath not set, try to determine the engineSourcePath by sky_engine setting.
         if (engineUri != null) {
-          engineSourcePath = fs.path.dirname(fs.path.dirname(fs.path.dirname(fs.path.dirname(engineUri.path))));
-          final bool dirExists = fs.isDirectorySync(fs.path.join(engineSourcePath, 'out'));
-          if (engineSourcePath == '/' || engineSourcePath.isEmpty || !dirExists)
+          engineSourcePath = fs.path.dirname(fs.path.dirname(fs.path.dirname(fs.path.dirname(fs.path.dirname(fs.path.dirname(engineUri.path))))));
+          if (engineSourcePath == '/' || engineSourcePath.isEmpty) {
```

Comment:
This might not work on Windows: if you get to the top of the filesystem, it might still contain the drive letter, or the UNC host for UNC paths.

Ideally, you should compare it to `fs.path.rootPrefix(engineUri.path)` instead, but it doesn't appear that FileSystem implements that method from dart:io.  Maybe you can check and see if `fs.path.dirname(engineSourcePath) == engineSourcePath`?

Reply:
This is a priorly existing logic and I will try to resolve.